src/FSharpVSPowerTools.Logic/Outlining/OutliningTagger.fs

```diff
@@ -0,0 +1,196 @@
+    seq {
+        yield mkFileIndexRange range.FileIndex componentInfo.Range.End range.End
+        match objectModel with
+        | ObjectModel(_, members, _) -> yield! Seq.collect visitSynMemberDefn members
+        | Simple _ -> yield! Seq.collect visitSynMemberDefn members
+    }
+
+let rec visitDeclaration (decl: SynModuleDecl) = 
+    seq {
+        match decl with
+        | SynModuleDecl.Let(_, bindings, _) ->
+            yield! Seq.map visitBinding bindings
+        | SynModuleDecl.Types(types, _) ->
+            yield! Seq.collect visitTypeDefn types
+        | SynModuleDecl.NestedModule(cmpInfo, decls, _, _) ->
+            yield mkFileIndexRange decl.Range.FileIndex cmpInfo.Range.End decl.Range.End
+            yield! Seq.collect visitDeclaration decls
+        | _ -> ()
+    }
+
+let visitModuleOrNamespace moduleOrNs =
+    seq {
+        let (SynModuleOrNamespace(_, _, decls, _, _, _, _)) = moduleOrNs
+        yield! Seq.collect visitDeclaration decls
+    }
+
+let visitAst tree =
+    seq {
+        match tree with
+        | ParsedInput.ImplFile(implFile) ->
+            let (ParsedImplFileInput(_, _, _, _, _, modules, _)) = implFile
+            yield! Seq.collect visitModuleOrNamespace modules
+        | _ -> ()
+    }
+
+type SnapshotPoint with
+    static member OfPos(snapshot: ITextSnapshot, p: pos) =
+        let (line, column) = Pos.toZ(p)
+        snapshot.GetLineFromLineNumber(line).Start.Add(column)
+
+type SnapshotSpan with
+    static member OfRange(snapshot: ITextSnapshot, r: range) =
+        SnapshotSpan(
+            SnapshotPoint.OfPos(snapshot, r.Start),
+            SnapshotPoint.OfPos(snapshot, r.End))
+
+type Tagger
+    ( buffer: ITextBuffer
+    , textDocument: ITextDocument
+    , serviceProvider : IServiceProvider
+    , projectFactory: ProjectFactory
+    , languageService: VSLanguageService) as self = 
+
+    let tagsChanged = Event<_,_>()
+    let mutable snapshotSpans : SnapshotSpan [] = [||]
+
+    let triggerUpdate newSnapshotSpans =
+        snapshotSpans <- newSnapshotSpans
+        tagsChanged.Trigger(
+            self,
+            SnapshotSpanEventArgs(
+                SnapshotSpan(
+                    buffer.CurrentSnapshot,
+                    0,
+                    buffer.CurrentSnapshot.Length - 1)))
+
+    let createTagSpan (ss: SnapshotSpan) =
+        try
+            let snapshot = ss.Snapshot
+            let firstLine = snapshot.GetLineFromPosition(ss.Start.Position)
+            let mutable lastLine = snapshot.GetLineFromPosition(ss.End.Position)
+
+            let nHintLines = lastLine.LineNumber - firstLine.LineNumber + 1
+            if nHintLines > MaxTooltipLines then
+                lastLine <- snapshot.GetLineFromLineNumber(firstLine.LineNumber + MaxTooltipLines - 1)
+
+            let missingLinesCount = Math.Max(nHintLines - MaxTooltipLines, 0)
+
+            let hintSnapshotSpan = SnapshotSpan(firstLine.Start, lastLine.End)
+            let hintText = lazy(
```

Comment:
`hintText` is used once later. Could you inline it at the call site or make `hintText` a function?

Making it `lazy` doesn't seem to have any advantage.